### PR TITLE
Add concatenate strategy merge corpora job

### LINKED ISSUE #146: MergeCorporaJob with MergeStrategy.FLAT produces invalid corpora if subcorpora are present
When using MergeStrategy.FLAT with corpus files structured as : 
corpus 
  subcorpus
    speaker-description*
    recording*

the result eliminates the `subcorpus` tags, doesn't copy the `speaker-description`s and loses the attribute name of subcorpus.
This results in a corpus file that is not compatible with the source files.

After discussing with @curufinwe I'm going to add a new strategy MergeStrategy.CONCATENATE that concatenates the subcorpus tags so that everything stays consistent with the source.

### PR DESCRIPTION
**Pull request by mattiadg**
Fix #146

### CODE REVIEW

File: lib/corpus.py [hunk 1]
```diff
@@ -170,6 +170,10 @@ def segments(self):
         for sc in self.subcorpora:
             yield from sc.segments()
 
+    def all_subcorpora(self):
```
Comment (curufinwe) [inline lib/corpus.py:173]: this would be `top_level_subcorpora`, not `subcorpora`(as subcorpora can be nested).

**Review by curufinwe — CHANGES_REQUESTED** [review]

File: lib/corpus.py [hunk 2]
```diff
@@ -180,6 +180,12 @@ def all_speakers(self):
         for sc in self.subcorpora:
             yield from sc.all_speakers()
 
+    def top_level_recordings(self):
+        yield from self.recordings
+
+    def top_level_subcorpora(self):
+        yield from self.subcorpora
+
```
Comment (michelwi) [inline lib/corpus.py:188]: Suggested implementation for `top_level_speakers()` (see prev comment)
```suggestion

    def top_level_speakers(self):
        yield from self.speakers.values()
```

Only semi related to this PR, but if `self.speakers` is a dict, `yield from self.speakers` would give the keys (speaker name) and not the values (speaker object), as would be necessary, right?
This should be fixed in `all_speakers` as well.

File: corpus/transform.py [hunk 3]
```diff
@@ -345,6 +346,11 @@ def run(self):
                 for rec in c.all_recordings():
                     merged_corpus.add_recording(rec)
                 merged_corpus.speakers.update(c.speakers)
+            elif self.merge_strategy == MergeStrategy.CONCATENATE:
+                for subcorpus in c.top_level_subcorpora():
+                    merged_corpus.add_subcorpus(subcorpus)
+                for rec in c.top_level_recordings():
+                    merged_corpus.add_recording(rec)
```
Comment (michelwi) [inline corpus/transform.py:353]: Do we also have to add the top level speakers?
```suggestion
                    merged_corpus.add_recording(rec)
                for speaker in c.top_level_speakers():
                    merged_corpus.add_speaker(speaker)
```

(with a suitable implementation of `Corpus.top_level_speakers()`)

File: lib/corpus.py (hunk 2 above)
Comment (mattiadg) [reply lib/corpus.py:188]: Related to this question, why is `self.speakers` a dictionary, if the value also contain the key (`speaker.name`)?

Comment (mattiadg) [reply lib/corpus.py:188]: Maybe this should be discussed in a new issue and not here

**Comment by mattiadg** [comment]
@curufinwe do you see more problems here?

File: lib/corpus.py (hunk 2 above)
Comment (curufinwe) [reply lib/corpus.py:188]: `top_level_speakers` should behave the same as `all_speakers`. It's very counter-intuitive that one returns values and the other keys. And as `all_speakers` already exists the one that needs to be changed is `top_level_speakers`.

Comment (mattiadg) [reply lib/corpus.py:188]: Are you sure that `all_speakers` is not buggy as it is now?

Comment (michelwi) [reply lib/corpus.py:188]: I agree that `top_level_speakers` should behave the same as `all_speakers`. However, I think the current way how `all_speakers` works is wrong:

`all_speakers` returns the keys of `self.speakers` which are of type `str`. 
But `Corpus.add_speaker` expects a `Speaker` object (in fact there is an assertion), so the code
```
for speaker in c.all_speakers():
     merged_corpus.add_speaker(speaker)
```
will not work.

I therefore propose to check if `all_speakers` is already used (and if returning the key is expected) and if not change both functions to return the values instead.

Comment (curufinwe) [reply lib/corpus.py:188]: Looks like the `all_speakers` code was added here: https://github.com/rwth-i6/i6_core/pull/43 And I did not find a use in i6_core or the AppTek repo for it, so it's probably safe to change it.

**Comment by mattiadg** [comment]
Shall I also do it in this PR?

Il ven 24 set 2021, 18:36 Eugen Beck ***@***.***> ha scritto:

File: lib/corpus.py (hunk 2 above)
Comment (mattiadg) [reply lib/corpus.py:188]: ok, so I probably didn't realize that `self.speakers` is a dict, thus the implementation

**Review by curufinwe — APPROVED** [review]

**Pull request merged**